### DISCUSSION TITLE
How to destroy animated entities?

### DISCUSSION THREAD

**Discussion by ebee1** [Dreaming381/Latios-Framework#10] [Q&A]
Is there a special way you need to destroy an entity using Latios-Framework's Kinemation?
I've tried several of the standard approaches using several of the normal command buffers and the Latios-World command buffer.  But i keep getting a NativeParallelHashMap error.
More specifically, when I run it without burst compile on it tells me:

```
ArgumentException: Key: Entity(73:3) is not present in the NativeParallelHashMap.
Unity.Collections.NativeHashMap2[TKey,TValue].ThrowKeyNotPresent (TKey key) 

```
Here's the point in code the error pointins to inside the SkeletonMeshBindingReactiveSystem:

```
for (int i = 0; i < deadExposedSkeletons2.Length; i++)
                {
                    var entity = deadExposedSkeletons2[i];
                    int index  = cullingManager.skeletonToCullingIndexMap[entity];  //<--- this one at line # 825
                    cullingManager.indexFreeList.Add(index);
                    cullingManager.skeletonToCullingIndexMap.Remove(entity);
                    cullingManager.cullingIndexToSkeletonMap.Remove(index);
                    indicesToClear.Value.Set(index, true);
                } 

```

Here's my current code: 
```
   private EndSimulationEntityCommandBufferSystem _endSimulationECBSystem;
    protected override void OnCreate()
    {
        _endSimulationECBSystem = World.GetOrCreateSystem<EndSimulationEntityCommandBufferSystem>();
    }

    protected override void OnUpdate()
    {
        return;
        var ecb =_endSimulationECBSystem.CreateCommandBuffer().AsParallelWriter();

        Entities
            .ForEach((Entity entity, int entityInQueryIndex, in MobKillTag killTag ,in LocalToWorld localToWorld) =>
            {
               ecb.DestroyEntity(entityInQueryIndex, entity);
            }).Schedule();

        _endSimulationECBSystem.AddJobHandleForProducer(this.Dependency);
    }
```

**Comment by Dreaming381** [comment]
That's a Kinemation bug. Sorry about that! It only affects exposed skeletons, which is why I never noticed. (My internal project that does dynamic spawning and despawning uses optimized skeletons.)

To fix it, beginning on line 823 of SkeletonMeshBindingReactiveSystem.cs, you want the code to look like this or something similar:
```csharp
var entity = deadExposedSkeletons2[i];
if (!cullingManager.skeletonToCullingIndexMap.ContainsKey(entity))
    continue;
int index  = cullingManager.skeletonToCullingIndexMap[entity];
```

Thanks for reaching out! I'll be sure to fix this in the next patch!

**Reply by ebee1** [reply]
Hey thanks so much for the swift reply. 

That fixed the one bug, but now i'm getting a new error after destroying a few entity's. It's random how many each time, but often before the first 10, however it doesn't crash the game, just spams this message:

```
ArgumentException: BitArray invalid arguments: pos 6 (must be 0-5), numBits 1 (must be greater than 0).
Unity.Collections.LowLevel.Unsafe.UnsafeBitArray.CheckArgs (System.Int32 pos, System.Int32 numBits) (at Library/PackageCache/com.unity.collections@1.3.1/Unity.Collections/UnsafeBitArray.cs:605)
Unity.Collections.LowLevel.Unsafe.UnsafeBitArray.Set (System.Int32 pos, System.Boolean value) (at Library/PackageCache/com.unity.collections@1.3.1/Unity.Collections/UnsafeBitArray.cs:133)
Latios.Kinemation.Systems.SkeletonMeshBindingReactiveSystem+ProcessNewAndDeadExposedSkeletonsJob.Execute () (at Library/PackageCache/com.latios.latiosframework@0f7ebfdc62/Kinemation/Systems/SyncPoint/SkeletonMeshBindingReactiveSystem.cs:819)
Unity.Jobs.IJobExtensions+JobStruct`1[T].Execute (T& data, System.IntPtr additionalPtr, System.IntPtr bufferRangePatchData, Unity.Jobs.LowLevel.Unsafe.JobRanges& ranges, System.Int32 jobIndex) (at <4a31731933e0419ca5a995305014ad37>:0)
```

I think it has to do with it not adding a spot from the indicesToClear. But i don't know anything about UnsafeBitArray's. 
And if i do the same:
  if (!cullingManager.skeletonToCullingIndexMap.ContainsKey(entity))  continue; 
on that upper loop, it still happens. 
```
indicesToClear.Value = new UnsafeBitArray(cullingManager.maxIndex.Value, allocator);

                for (int i = 0; i < deadExposedSkeletons.Length; i++)
                {
                    var entity = deadExposedSkeletons[i];
                    int index  = cullingManager.skeletonToCullingIndexMap[entity];
                    cullingManager.indexFreeList.Add(index);
                    cullingManager.skeletonToCullingIndexMap.Remove(entity);
                    cullingManager.cullingIndexToSkeletonMap.Remove(index);
                    indicesToClear.Value.Set(index, true);                                 //<--------------------------------------- #819
                }
                for (int i = 0; i < deadExposedSkeletons2.Length; i++)
                {
                    var entity = deadExposedSkeletons2[i];
                    if (!cullingManager.skeletonToCullingIndexMap.ContainsKey(entity))
                    {
                        continue;
                    }

                    int index  = cullingManager.skeletonToCullingIndexMap[entity];
                    cullingManager.indexFreeList.Add(index);
                    cullingManager.skeletonToCullingIndexMap.Remove(entity);
                    cullingManager.cullingIndexToSkeletonMap.Remove(index);
                    indicesToClear.Value.Set(index, true);
                }
```

If it helps to have context, we are making an endless wave survival game, so we are constantly creating and destroying entity's.

**Reply by Dreaming381** [reply]
Once again, sorry. You ran into a not-very-well-tested codepath.

The first of the loops does not need the if check. The reason the second loop needs it is that there are two different EntityQueries which detect destroyed entities. And your destroyed entities happened to match both, so the latter loop was trying to free an entity the first loop already handled.

As for this new bug, it is an off-by-one bug. On the line that creates the new UnsafeBitArray (810), it needs to allocate a size of `cullingManager.maxIndex.Value + 1`.

**Reply by ebee1** [reply]
Thanks so much that fixed it. And thanks for the whole latios gramework. It's staggeringly impressive and awesome that you're sharing it.

**Reply by Dreaming381** [reply]
Absolutely! Thanks for reaching out and providing good details and stack traces.

I look forward to seeing what you build with it!
